[WIP] Remove leading backslash for locateClassFile()

VolCh opened the pull request:
In some cases (maybe only in the global namespace) function get class name with leading backslash ('\SomeClassName'). Neither class_exists(), nor ComposerLocator::locateClass (with classmap) can't find file with this class. This commit fix it.

Closes #80

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/48203bec-5c67-41c1-af95-31a2a99ebd79) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/48203bec-5c67-41c1-af95-31a2a99ebd79) completed: **No new issues**

aik099 commented:
Would you please also add a test, that covers fixed behavior?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/1481953c-f9b4-4d4e-8215-c037c6eb5d2a) was created.

VolCh commented:
Test added. I decided to move fix in ComposerLoader because exactly composer expects FQCN without leading slash. I'm not sure about contract of CallableLocator & LocatorInterface in all.

aik099 commented:
Questions:

1. So you're having classes in a PHP file that doesn't have `namespace ...` declaration, right?
2. are you accessing classes in a file through the `\Go\ParserReflection\ReflectionFileNamespace` class instance or directly by calling `\Go\ParserReflection\ReflectionEngine::parseClass` method?

Observations:

1. As I can see the `\Go\ParserReflection\ReflectionFileNamespaceTest` test doesn't cover such behavior.

VolCh commented:
1. Yes. Project is a mix of  legacy (from PHP4) and modern code
2. We use Codeception/AspectMock and got trace 
```
() at /home/user/app/vendor/goaop/parser-reflection/src/ReflectionEngine.php:125
 Go\ParserReflection\ReflectionEngine::locateClassFile() at /home/user/app/vendor/goaop/parser-reflection/src/ReflectionEngine.php:140
 Go\ParserReflection\ReflectionEngine::parseClass() at /home/user/app/vendor/goaop/parser-reflection/src/ReflectionClass.php:44
 Go\ParserReflection\ReflectionClass->__construct() at /home/user/app/vendor/goaop/parser-reflection/src/ValueResolver/NodeExpressionResolver.php:478
 Go\ParserReflection\ValueResolver\NodeExpressionResolver->fetchReflectionClass() at /home/user/app/vendor/goaop/parser-reflection/src/ValueResolver/NodeExpressionResolver.php:271
 Go\ParserReflection\ValueResolver\NodeExpressionResolver->resolveExprClassConstFetch() at /home/user/app/vendor/goaop/parser-reflection/src/ValueResolver/NodeExpressionResolver.php:117
 Go\ParserReflection\ValueResolver\NodeExpressionResolver->resolve() at /home/user/app/vendor/goaop/parser-reflection/src/ValueResolver/NodeExpressionResolver.php:99
 Go\ParserReflection\ValueResolver\NodeExpressionResolver->process() at /home/user/app/vendor/goaop/parser-reflection/src/ReflectionParameter.php:100
 Go\ParserReflection\ReflectionParameter->__construct() at /home/user/app/vendor/goaop/parser-reflection/src/Traits/ReflectionFunctionLikeTrait.php:164
 Go\ParserReflection\ReflectionMethod->getParameters() at /home/user/app/vendor/codeception/aspect-mock/src/AspectMock/Intercept/BeforeMockTransformer.php:56
 AspectMock\Intercept\BeforeMockTransformer->transform() at /home/user/app/vendor/goaop/framework/src/Instrument/Transformer/CachingTransformer.php:124
...
```

We have LocatorInterface::locateClass($className). I'm not sure about its contract. What kind of class names it should locate? Is leading slashes bug a composer specific behavior?

aik099 commented:
You're trace appears to be incomplete, because I'm not seeing what searches for classes in a file. From trace fragment you've posted I can see that a function (or class method) is being analyzed and one of parameters is type-hinted with class without namespace.

VolCh commented:
My fault, full trace:

 ```
Exception trace:
 () at /home/user/app/vendor/goaop/parser-reflection/src/ReflectionEngine.php:125
 Go\ParserReflection\ReflectionEngine::locateClassFile() at /home/user/app/vendor/goaop/parser-reflection/src/ReflectionEngine.php:140
 Go\ParserReflection\ReflectionEngine::parseClass() at /home/user/app/vendor/goaop/parser-reflection/src/ReflectionClass.php:44
 Go\ParserReflection\ReflectionClass->__construct() at /home/user/app/vendor/goaop/parser-reflection/src/ValueResolver/NodeExpressionResolver.php:478
 Go\ParserReflection\ValueResolver\NodeExpressionResolver->fetchReflectionClass() at /home/user/app/vendor/goaop/parser-reflection/src/ValueResolver/NodeExpressionResolver.php:271
 Go\ParserReflection\ValueResolver\NodeExpressionResolver->resolveExprClassConstFetch() at /home/user/app/vendor/goaop/parser-reflection/src/ValueResolver/NodeExpressionResolver.php:117
 Go\ParserReflection\ValueResolver\NodeExpressionResolver->resolve() at /home/user/app/vendor/goaop/parser-reflection/src/ValueResolver/NodeExpressionResolver.php:99
 Go\ParserReflection\ValueResolver\NodeExpressionResolver->process() at /home/user/app/vendor/goaop/parser-reflection/src/ReflectionParameter.php:100
 Go\ParserReflection\ReflectionParameter->__construct() at /home/user/app/vendor/goaop/parser-reflection/src/Traits/ReflectionFunctionLikeTrait.php:164
 Go\ParserReflection\ReflectionMethod->getParameters() at /home/user/app/vendor/codeception/aspect-mock/src/AspectMock/Intercept/BeforeMockTransformer.php:56
 AspectMock\Intercept\BeforeMockTransformer->transform() at /home/user/app/vendor/goaop/framework/src/Instrument/Transformer/CachingTransformer.php:124
 Go\Instrument\Transformer\CachingTransformer->processTransformers() at /home/user/app/vendor/goaop/framework/src/Instrument/Transformer/CachingTransformer.php:64
 Go\Instrument\Transformer\CachingTransformer->transform() at /home/user/app/vendor/goaop/framework/src/Instrument/ClassLoading/SourceTransformingLoader.php:134
 Go\Instrument\ClassLoading\SourceTransformingLoader::transformCode() at /home/user/app/vendor/goaop/framework/src/Instrument/ClassLoading/SourceTransformingLoader.php:101
 Go\Instrument\ClassLoading\SourceTransformingLoader->filter() at /home/user/app/vendor/goaop/framework/src/Instrument/ClassLoading/AopComposerLoader.php:137
 Go\Instrument\ClassLoading\AopComposerLoader->loadClass() at /home/user/app/vendor/goaop/framework/src/Instrument/ClassLoading/AopComposerLoader.php:137
 Go\Instrument\ClassLoading\AopComposerLoader->loadClass() at n/a:n/a
 spl_autoload_call() at /home/user/app/tests/global.php:10
 include_once() at /home/user/app/tests/unit/_bootstrap.php:5
 require_once() at /home/user/app/vendor/codeception/codeception/src/Codeception/Subscriber/Bootstrap.php:34
 Codeception\Subscriber\Bootstrap->loadBootstrap() at n/a:n/a
 call_user_func() at /home/user/app/vendor/symfony/event-dispatcher/EventDispatcher.php:212
 Symfony\Component\EventDispatcher\EventDispatcher->doDispatch() at /home/user/app/vendor/symfony/event-dispatcher/EventDispatcher.php:44
 Symfony\Component\EventDispatcher\EventDispatcher->dispatch() at /home/user/app/vendor/codeception/codeception/src/Codeception/SuiteManager.php:88
 Codeception\SuiteManager->initialize() at /home/user/app/vendor/codeception/codeception/src/Codeception/Codecept.php:186
 Codeception\Codecept->runSuite() at /home/user/app/vendor/codeception/codeception/src/Codeception/Codecept.php:157
 Codeception\Codecept->run() at /home/user/app/vendor/codeception/codeception/src/Codeception/Command/Run.php:341
 Codeception\Command\Run->execute() at /home/user/app/vendor/symfony/console/Command/Command.php:264
 Symfony\Component\Console\Command\Command->run() at /home/user/app/vendor/symfony/console/Application.php:888
 Symfony\Component\Console\Application->doRunCommand() at /home/user/app/vendor/symfony/console/Application.php:224
 Symfony\Component\Console\Application->doRun() at /home/user/app/vendor/symfony/console/Application.php:125
 Symfony\Component\Console\Application->run() at /home/user/app/vendor/codeception/codeception/src/Codeception/Application.php:103
 Codeception\Application->run() at /home/user/app/vendor/codeception/codeception/codecept:36
 require_once() at /home/user/app/bin/codecept:8
```


The problem class never used for type-hintnig, but in the method declaration as default param value `public function someMethod($someParam = ProblemClass::SOME_CONST ) `.

aik099 commented:
Something is terribly wrong here. The first thing the Composer class loader (`\Composer\Autoload\ClassLoader::findFile`) does it actually trimming leading slash:

```php
// work around for PHP 5.3.0 - 5.3.2 https://bugs.php.net/50731
if ('\\' == $class[0]) {
    $class = substr($class, 1);
}
```
so what you're doing in this PR has no effect unless you're using older Composer version without this fix and are on PHP 5.3.0 till 5.3.2. Is that the case?

VolCh commented:
No, we use PHP 5.6 & 7 and last composer. This workaround was removed about year ago with  https://github.com/composer/composer/commit/2853e82400a113620e05504c52729c61b5a53f77#diff-02d8fd4e825f60819515851876a3e6e6

Composer expect exactly class name without leading slash (in `\ClassName::class` PHP format), at least when it use class map.

Probably in fact my PR is workaround for some bug(s) some where else in the goaop/parser-reflection or goaop/framework or some where else in our stack, and `ReflectionEngine::locateClassFile()` should never get class name with leading slash but it get in our case. Probably this PR  is  https://github.com/goaop/parser-reflection/pull/68#issuecomment-277447476 & https://github.com/goaop/parser-reflection/pull/68#issuecomment-277451420 related.

I'm sure that `\Composer\Autoload\ClassLoader::findFile` expects  class name in `\ClassName::class` PHP format now, without leading slash like PHP autoload engine:
```
spl_autoload_register(function ($className) { var_dump($className); });
new \ClassName();
```
output is `string(9) "ClassName"`. But I'm not sure about `LocatorInterface::locateClass()` expectations. Should it never get leading backslashes in class name or should implementations work in both, with and without slashes, cases. Or, may be, only with leading. The phpdoc of `LocatorInterface::locateClass()` does not explain class name format in `string $className Name of the class`

aik099 commented:
The `Go\ParserReflection\ReflectionEngine::parseClass` method is the only one, that triggers a locator. Can you please see:

1. what argument it's receiving when a namespaced VS non-namespaced class comes in
2. in library's own test suite which argument it receives
3. if you scan single file using `\Go\ParserReflection\ReflectionFileNamespace` class, then what arguments will `Go\ParserReflection\ReflectionEngine::parseClass` method receive when attempting to get classes from that file

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/561761aa-542b-414d-9925-3f0af413309e) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/561761aa-542b-414d-9925-3f0af413309e) completed: **2 new issues, 1 updated code elements**

VolCh commented:
I pushed a "for play" (not for merge!) commit with test which shows problem behavior. I can't reproduce it with this repo way in the reasonable time. Possible it helps you to understand isuue.

lisachenko commented:
Hi, what is the current status of this PR?

aik099 commented:
Not ready for merging. We can't create a test case to reproduce the problem. Theoretically if during class parsing (the `OneClass` in example below) a method parameter has another class (the `TwoClass` in example below) constant as default value and that other class isn't autoloaded yet AND isn't namespaced, then class locator is receiving class name with leasing `\` and then Composer isn't able to auto-load actual class.

```php
class OneClass {
    protected function methodOne($arg = TwoClass::CONSTANT) {}
}
```

We're unsure if the problem is with `goaop/parser-reflection` or the `Codeception/AspectMock`, that @VolCh is using.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/aca7d760-1de7-42b1-8d54-64dce069d7d9) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/aca7d760-1de7-42b1-8d54-64dce069d7d9) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/ebbc2519-ecee-40a2-96ec-a8c63b52e953) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/ebbc2519-ecee-40a2-96ec-a8c63b52e953) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/241e3a65-9715-431c-ba75-71351de4e3bf) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/1afbaa6a-7ffb-49e6-b3af-9ad8260af8a6) was created.

VolCh commented:
I created pure test that shows inconsistent behavior of parser in cases of `GlobalNsClass::A` and `\GlobalNsClass::A` in the default param value. ReflectionEngine::getClass() got `\GlobalNsClass` and `GlobalNsClass` respectively  (sic! the presence of leading slash inverted)

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/1afbaa6a-7ffb-49e6-b3af-9ad8260af8a6) completed: **No new issues**

aik099 commented on tests/ReflectionParameterTest.php:
```diff
@@ -171,6 +171,21 @@ public function testParamBuiltInClassConst()
+        $parsedFunction  = $parsedClass->getMethod('foo');
+
+        $parameters = $parsedFunction->getParameters();
+        $this->assertTrue($parameters[0]->isDefaultValueConstant());
```
Please specify message to each of the `assert...` method calls in this test, so that we can see which of two assertions failed based on the error message, because they're comparing the same things now.

VolCh commented:
- Removed passed assertions for isDefaultValueConstant()
- Added assertions for namespaced class in three cases (with slash, without slash and with use)

It seems like parser doesn't understand that in global namespace the classname is FQCN in both cases with & without leading slash.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/0c2b8c5e-ea8a-45a2-84d7-ee8eb749cdcc) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/0c2b8c5e-ea8a-45a2-84d7-ee8eb749cdcc) completed: **No new issues**

aik099 commented:
For the sake of better reporting (right now 1st failed assertion prevents other assertions to be made) I recommend:

1. splitting one fixture function with 4 parameters into 4 functions with 1 parameter each
2. using same name for parameter in each of 4 functions
3. moving function names and expected class name into data provider
4. using names in data provider array and that name will later be shown automatically (no need to specify assertion message now)

This way each failure would be seen easily.

aik099 commented:
We must also figure out who is responsible for loosing the `\`:

* this library
* or the php-parser

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/08722640-d959-4f04-a246-94061177d9f8) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/08722640-d959-4f04-a246-94061177d9f8) completed: **No new issues**

VolCh commented:
I rewrited tests with data provider

aik099 commented:
Nice.

What is expected notation as a result of this PR:

1. always have leading `\`
2. always remove leading `\`
3. keep leading `\` as was written in code

?

If we'll use `ReflectionFile` class (https://github.com/goaop/parser-reflection/blob/master/src/ReflectionFile.php) to:

1. find namespaces in a file (one file with without namespace and another one with namespace)
2. find classes in each namespace
3. show the FQCN of these classes

Then will leading `\` be present in any of these? Answer to this question will suggest expected behavior in question 1 above.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/a8c80602-042c-4b26-81ea-43c786b33560) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/a8c80602-042c-4b26-81ea-43c786b33560) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/9e088f1d-32ba-497e-bbce-500e043fc483) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/9e088f1d-32ba-497e-bbce-500e043fc483) completed: **No new issues**

VolCh commented:
@pchabbert After discussion with @aik099 and https://github.com/goaop/parser-reflection/pull/89 I not sure that this PR  fixes my problem correctly. On other hands, it prevents any related issues in the feature calling internal locator in the manner of composer locator. Or we may trim backslash in ComposerLocator and document LocatorInterface::locateClass() with note that implementations must be leading backslash agnostic.

VolCh commented:
outdated after last #89 changes
